File: docs/snippets/all/archetypes/segmentation_image_simple.rs
//! Create and log a segmentation image.

use ndarray::{Array, ShapeBuilder as _, s};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let rec = rerun::RecordingStreamBuilder::new("rerun_example_segmentation_image").spawn()?;

    // create a segmentation image
    let mut image = Array::<u8, _>::zeros((8, 12).f());
    image.slice_mut(s![0..4, 0..6]).fill(1);
    image.slice_mut(s![4..8, 6..12]).fill(2);

    // create an annotation context to describe the classes
    let annotation = rerun::AnnotationContext::new([
        (1, "red", rerun::Rgba32::from_rgb(255, 0, 0)),
        (2, "green", rerun::Rgba32::from_rgb(0, 255, 0)),
    ]);

    // log the annotation and the image
    rec.log_static("/", &annotation)?;

    rec.log("image", &rerun::SegmentationImage::try_from(image)?)?;

    Ok(())
}

File: docs/snippets/all/archetypes/segmentation_image_simple.cpp
// Create and log a segmentation image.

#include <rerun.hpp>

#include <algorithm> // std::fill_n
#include <vector>

int main(int argc, char* argv[]) {
    const auto rec = rerun::RecordingStream("rerun_example_segmentation_image");
    rec.spawn().exit_on_failure();

    // Create a segmentation image
    const int HEIGHT = 8;
    const int WIDTH = 12;
    std::vector<uint8_t> data(WIDTH * HEIGHT, 0);
    for (auto y = 0; y < 4; ++y) {                                         // top half
        std::fill_n(data.begin() + y * WIDTH, 6, static_cast<uint8_t>(1)); // left half
    }
    for (auto y = 4; y < 8; ++y) {                                             // bottom half
        std::fill_n(data.begin() + y * WIDTH + 6, 6, static_cast<uint8_t>(2)); // right half
    }

    // create an annotation context to describe the classes
    rec.log_static(
        "/",
        rerun::AnnotationContext({
            rerun::AnnotationInfo(1, "red", rerun::Rgba32(255, 0, 0)),
            rerun::AnnotationInfo(2, "green", rerun::Rgba32(0, 255, 0)),
        })
    );

    rec.log("image", rerun::SegmentationImage(data.data(), {WIDTH, HEIGHT}));
}
